In [1]:
# read in libraries
import pandas as pd
import numpy as np
import pickle
pd.options.mode.chained_assignment = None 
pd.errors.DtypeWarning = None

In [2]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from urllib.request import urlopen
from bs4 import BeautifulSoup
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
from scipy import stats
from scipy.stats import ttest_ind
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import VotingClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import GridSearchCV
from sklearn.pipeline import Pipeline
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import f1_score

In [ ]:
# read in datasets

# kaggle datasets
games = pd.read_csv('games.csv') 
games_details = pd.read_csv('games_details.csv')
players = pd.read_csv('players.csv')
ranking = pd.read_csv('ranking.csv')
teams = pd.read_csv('teams.csv')

# datasets created from internet
jersey = pd.read_csv('jersey2.csv')
twok = pd.read_csv('2kRank.csv')
pacedata = {'season':[2012,2013,2014,2015,2016,2017,2018,2019,2020],
           'pace':[92.0,93.9,93.9,95.8,96.4,97.3,100.0,100.3,99.2]}
pace = pd.DataFrame(pacedata)

# load in scraped all star info
# get this from ASG Scrape
all_star_appearances = pickle.load(open('all_star_appearances.pickle', 'rb'))

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (6) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [ ]:
# join date to individual games stats dataset
games_details2 = pd.merge(games_details, games[['GAME_DATE_EST', 'GAME_ID']], on = 'GAME_ID', how = 'left')

# extract year and month for every row
games_details2['year'] =  pd.to_numeric(games_details2['GAME_DATE_EST'].apply(lambda x: x.split('-')[0]))
games_details2['month'] =  pd.to_numeric(games_details2['GAME_DATE_EST'].apply(lambda x: x.split('-')[1]))

# label each row with season based on year and month
games_details2['season'] = games_details2[['year','month']].apply(lambda x: x['year']-1 if (x['month']==1) else x['year'], axis=1)

# keep only game info for September through January in the 2012-2020 seasons
games_details2 = games_details2[games_details2['month'].isin([9,10,11,12,1])]
games_details3 = games_details2[(games_details2['season']<2021) & (games_details2['season']>2011)]

# get rid of 2020 finals
games_details3 = games_details3[~((games_details3['season']==2020) & (games_details3['month'].isin([9,10])))]

# drop duplicates
games_details3 = games_details3.drop_duplicates(subset=['GAME_ID','PLAYER_ID'])

In [ ]:
# get rid of rows with comments (means they did not play)
games_details3 = games_details3[games_details3['COMMENT'].isnull()]

# get rid of one erroneous row
games_details3 = games_details3[games_details3['PTS'].notnull()]

# extract number of minutes played
games_details3['minutes'] =  pd.to_numeric(games_details3['MIN'].apply(lambda x: x.split(':')[0]))

# add a column to indicate they played in that game (useful for aggregation in next step)
games_details3['GP'] = 1

In [ ]:
# uses groupby to summarize by player and season
grouped = games_details3.groupby(['PLAYER_ID','season']).apply(lambda s: pd.Series({ 
    "PLAYER_NAME": s["PLAYER_NAME"].mode()[0],
    "TEAM_ABBREVIATION": s["TEAM_ABBREVIATION"].mode()[0],
    "TEAM_ID": s["TEAM_ID"].mode()[0],
    "GP": s["GP"].sum(),
    "Minutes": s["minutes"].mean(),
    "FGM": s["FGM"].mean(),
    "FGA": s["FGA"].mean(),
    "FG3M": s["FG3M"].mean(),
    "FG3A": s["FG3A"].mean(),
    "FTM": s["FTM"].mean(),
    "FTA": s["FTA"].mean(),
    "OREB": s["OREB"].mean(),
    "DREB": s["DREB"].mean(),
    "AST": s["AST"].mean(),
    "STL": s["STL"].mean(),
    "BLK": s["BLK"].mean(),
    "TO": s["TO"].mean(), 
    "PF": s["PF"].mean(),
    "PTS": s["PTS"].mean(),
    "PLUS_MINUS": s["PLUS_MINUS"].mean(),
    "Starting_Position": s["START_POSITION"].mode(),
})).reset_index()

In [ ]:
# Fix Starting_Position column
grouped['Starting_Position'] = grouped['Starting_Position'].apply(lambda x: "None" if len(x)==0 else x[0])

In [ ]:
# merge pace data
grouped = pd.merge(grouped, pace, on = ['season'], how = 'left')

# divide pace data by 100
grouped['pace'] = grouped['pace']/100

# divide relevant statistics by pace/100
grouped[['FGM','FGA','FG3M','FG3A','FTM','FTA','OREB',
       'DREB', 'AST', 'STL', 'BLK', 'TO', 'PF', 'PTS', 'PLUS_MINUS']] = grouped[['FGM','FGA','FG3M','FG3A','FTM','FTA','OREB',
       'DREB', 'AST', 'STL', 'BLK', 'TO', 'PF', 'PTS', 'PLUS_MINUS']].div(grouped.pace, axis=0)

In [ ]:
# PER
grouped['PER'] = (grouped['FGM']*85.910 + grouped['STL']*53.897 + 
grouped['FG3M']*51.757 + grouped['FTM']*46.845 + grouped['BLK']*39.190
+ grouped['OREB']*39.190+ grouped['AST']*34.677 + grouped['DREB']*14.707
- grouped['PF']*17.174 - (grouped['FTA']-grouped['FTM'])*20.091 - (grouped['FGA']-grouped['FGM'])*39.190
- grouped['TO']*53.897)*(1/grouped['Minutes'])

# True Shooting Percentage
grouped['TSP'] = grouped['PTS']/(2*(grouped['FGA']+.44*grouped['FTA']))

In [ ]:
# keep only January 20 data
ranking2 = ranking[(ranking['STANDINGSDATE'].apply(lambda x: x.split('-')[1])=='01') & (ranking['STANDINGSDATE'].apply(lambda x: x.split('-')[2])=='20')]

# get season from the year
ranking2['season'] = ranking2['SEASON_ID']-20000
grouped = pd.merge(grouped, ranking2[['TEAM_ID', 'season','CONFERENCE','W_PCT']], on = ['TEAM_ID','season'], how = 'left')

In [ ]:
# merge jersey data
grouped = pd.merge(grouped, jersey, on = ['season','PLAYER_ID'], how = 'left')

In [ ]:
# merge 2k data
grouped = pd.merge(grouped, twok[['PLAYER_ID','season','2KRank']],on=['PLAYER_ID','season'],how='left')

In [ ]:
# function that defines if a player was an All star the previous year
def was_AS_last_year(row):
    if row['season'] == 1999:
        return 1 if 1998 in all_star_appearances[row['PLAYER_NAME']] else 0
    return 1 if row['season'] in all_star_appearances[row['PLAYER_NAME']] else 0

# adds if a player was an ALl star the previous year
grouped['LastASG?'] = grouped[['PLAYER_NAME', 'season']].apply(was_AS_last_year, axis=1)

# adds the total amount of prior selections a player had as of that year (not including that year)
grouped['PriorASG'] = grouped[['PLAYER_NAME', 'season']].apply(lambda row : sum(y<=row['season'] for y in all_star_appearances[row['PLAYER_NAME']]), axis=1)

# adds wether a player was selected as an all star that year
grouped['Selected?'] = grouped[['PLAYER_NAME', 'season']].apply(lambda row : 1 if row['season']+1 in all_star_appearances[row['PLAYER_NAME']] else 0, axis=1)


In [ ]:
# get total games played by each team up to that point in each season
game_count1 = games_details3.groupby(['season','GAME_ID','TEAM_ID']).GP.mean().reset_index()
game_count2 = game_count1.groupby(['season','TEAM_ID']).GP.sum().reset_index()

# merges Games Played rate with main dataset
grouped = pd.merge(grouped,game_count2,on=['season','TEAM_ID'],how='left')

# calculates percentage games played
grouped['GP_Per'] = grouped['GP_x'] / grouped['GP_y']

In [ ]:
#fill null values
grouped['PLUS_MINUS'] = grouped['PLUS_MINUS'].fillna(0)
grouped['TSP'] = grouped['TSP'].fillna(0)
grouped['PER'] = grouped['PER'].fillna(0)
grouped['2KRank'] = grouped['2KRank'].fillna(500)

In [ ]:
grouped.isnull().sum(axis = 0)


PLAYER_ID            0
season               0
PLAYER_NAME          0
TEAM_ABBREVIATION    0
TEAM_ID              0
GP_x                 0
Minutes              0
FGM                  0
FGA                  0
FG3M                 0
FG3A                 0
FTM                  0
FTA                  0
OREB                 0
DREB                 0
AST                  0
STL                  0
BLK                  0
TO                   0
PF                   0
PTS                  0
PLUS_MINUS           0
Starting_Position    0
pace                 0
PER                  0
TSP                  0
CONFERENCE           0
W_PCT                0
Jersey               0
2KRank               0
LastASG?             0
PriorASG             0
Selected?            0
GP_y                 0
GP_Per               0
dtype: int64

In [ ]:
grouped.loc[(grouped['PLAYER_NAME']=='Dwyane Wade') & (grouped['season']==2018),'Selected?'] = 0
grouped.loc[(grouped['PLAYER_NAME']=='Dirk Nowitzki') & (grouped['season']==2018),'Selected?'] = 0

In [ ]:
grouped

PLAYER_ID  season        PLAYER_NAME  ... Selected?  GP_y    GP_Per
0            255    2012         Grant Hill  ...         0    55  0.218182
1            467    2012         Jason Kidd  ...         0    49  0.877551
2            703    2012        Kurt Thomas  ...         0    49  0.673469
3            708    2012      Kevin Garnett  ...         1    51  0.960784
4            708    2013      Kevin Garnett  ...         0    51  0.843137
...          ...     ...                ...  ...       ...   ...       ...
5011     1630270    2020       Xavier Sneed  ...         0    24  0.125000
5012     1630271    2020     Brodric Thomas  ...         0    22  0.318182
5013     1630273    2020  Freddie Gillespie  ...         0    23  0.043478
5014     1630278    2020         Ade Murkey  ...         0    22  0.045455
5015  1962937827    2017          Matt Matt  ...         0    56  0.017857

[5016 rows x 35 columns]

In [ ]:
grouped.columns


Index(['PLAYER_ID', 'season', 'PLAYER_NAME', 'TEAM_ABBREVIATION', 'TEAM_ID',
       'GP_x', 'Minutes', 'FGM', 'FGA', 'FG3M', 'FG3A', 'FTM', 'FTA', 'OREB',
       'DREB', 'AST', 'STL', 'BLK', 'TO', 'PF', 'PTS', 'PLUS_MINUS',
       'Starting_Position', 'pace', 'PER', 'TSP', 'CONFERENCE', 'W_PCT',
       'Jersey', '2KRank', 'LastASG?', 'PriorASG', 'Selected?', 'GP_y',
       'GP_Per'],
      dtype='object')

In [ ]:
df = grouped[['PLAYER_ID', 'season', 'PLAYER_NAME', 'TEAM_ABBREVIATION', 'TEAM_ID',
       'GP_x', 'Minutes', 'FGM', 'FGA', 'FG3M', 'FG3A', 'FTM', 'FTA', 'OREB',
       'DREB', 'AST', 'STL', 'BLK', 'TO', 'PF', 'PTS', 'PLUS_MINUS', 'PER', 'TSP',
       'W_PCT', 'Jersey', '2KRank', 'LastASG?', 'PriorASG', 'Selected?', 'Starting_Position',
       'pace', 'CONFERENCE', 'GP_y', 'GP_Per']]

In [ ]:
df

PLAYER_ID  season        PLAYER_NAME  ... CONFERENCE  GP_y    GP_Per
0            255    2012         Grant Hill  ...       West    55  0.218182
1            467    2012         Jason Kidd  ...       East    49  0.877551
2            703    2012        Kurt Thomas  ...       East    49  0.673469
3            708    2012      Kevin Garnett  ...       East    51  0.960784
4            708    2013      Kevin Garnett  ...       East    51  0.843137
...          ...     ...                ...  ...        ...   ...       ...
5011     1630270    2020       Xavier Sneed  ...       East    24  0.125000
5012     1630271    2020     Brodric Thomas  ...       West    22  0.318182
5013     1630273    2020  Freddie Gillespie  ...       West    23  0.043478
5014     1630278    2020         Ade Murkey  ...       West    22  0.045455
5015  1962937827    2017          Matt Matt  ...       West    56  0.017857

[5016 rows x 35 columns]

In [ ]:
# replace nan with 0
df.fillna(0, inplace = True)

In [ ]:
# Create a list with numeric importatn stats
xcols = df.columns[6 : 29].to_list()

print(xcols)

['Minutes', 'FGM', 'FGA', 'FG3M', 'FG3A', 'FTM', 'FTA', 'OREB', 'DREB', 'AST', 'STL', 'BLK', 'TO', 'PF', 'PTS', 'PLUS_MINUS', 'PER', 'TSP', 'W_PCT', 'Jersey', '2KRank', 'LastASG?', 'PriorASG']


In [ ]:
# set train and test up
X_train, X_test, y_train, y_test = train_test_split(df[xcols], df['Selected?'], 
                                                    train_size = 0.8, random_state = 1)
print('training data:', X_train.shape)
print('test data:', X_test.shape)


training data: (4012, 23)
test data: (1004, 23)


In [ ]:
X_test.replace([np.inf, -np.inf], np.nan, inplace=True)
X_test.fillna(0, inplace=True)
X_train.replace([np.inf, -np.inf], np.nan, inplace=True)
X_train.fillna(0, inplace=True)

In [ ]:
# logistic regression
# even without solver specified, same results are shown
log_reg = LogisticRegression(solver = 'lbfgs', max_iter = 3000)

# Fit the model to the training data
clf = log_reg.fit(X_train, y_train)

# Get accuracy stats
print('training accuracy: {}'.format(clf.score(X_train, y_train).round(3)))
print('test accuracy: {}'.format(clf.score(X_test, y_test).round(3)))

training accuracy: 0.982
test accuracy: 0.985


In [ ]:
# Put the coefficients into a new dataframe
coef = pd.concat([pd.DataFrame(xcols),
                  pd.DataFrame(np.transpose(clf.coef_))], 
                 axis = 1)

coef.columns = ['feature','coefficient']

coef.sort_values(by=['coefficient'], 
                 ascending = False, 
                 inplace = True)

# Examine the features/stats with the 10 largest positive and negative coefficients
print('Ten largest positive features:\n', coef.head(10), '\n')
print('Ten largest negative features:\n', coef.tail(10))


Ten largest positive features:
      feature  coefficient
18     W_PCT     1.849673
11       BLK     0.715427
14       PTS     0.681582
21  LastASG?     0.465860
19    Jersey     0.404852
9        AST     0.393917
10       STL     0.364938
8       DREB     0.327467
7       OREB     0.293951
3       FG3M     0.202758 

Ten largest negative features:
     feature  coefficient
0   Minutes     0.142660
5       FTM     0.137352
17      TSP     0.110844
20   2KRank    -0.010645
12       TO    -0.052489
6       FTA    -0.141825
4      FG3A    -0.142887
16      PER    -0.146543
2       FGA    -0.385090
13       PF    -0.507503


In [ ]:
# Adding the features with the 10 largest positive and negative coefficients 
# to a new list for training purposes

xcols2 = coef.feature[0:10].to_list()
xcols2 += coef.feature[-10:].to_list()
print(xcols2)

['W_PCT', 'BLK', 'PTS', 'LastASG?', 'Jersey', 'AST', 'STL', 'DREB', 'OREB', 'FG3M', 'Minutes', 'FTM', 'TSP', '2KRank', 'TO', 'FTA', 'FG3A', 'PER', 'FGA', 'PF']


In [ ]:
# test train fpr large coefficients
X_train, X_test, y_train, y_test = train_test_split(df[xcols2], 
                                                    df['Selected?'], 
                                                    train_size = 0.8, 
                                                    random_state = 1)
print('training data:', X_train.shape)
print('test data:', X_test.shape)

# fit the model to the training data
clf2 = log_reg.fit(X_train, y_train)

# get accuracy stats
print('training accuracy: {}'.format(clf2.score(X_train, y_train).round(3)))
print('test accuracy: {}'.format(clf2.score(X_test, y_test).round(3)))

training data: (4012, 20)
test data: (1004, 20)
training accuracy: 0.982
test accuracy: 0.988


In [ ]:
# Regularization with logistic regression & C parameter with logistic regression

# high C value means trust the data a lot; low means ehhh
cset = [.001, .01, .1, 1, 10]

for i in cset:
    print('C =', i)
    log_reg = LogisticRegression(solver = 'lbfgs', 
                                 max_iter = 3000, 
                                 C = i)
    
    clf = log_reg.fit(X_train, y_train)
    
    print('training accuracy: {}'.format(clf.score(X_train, y_train).round(3)))
    print('test accuracy: {}'.format(clf.score(X_test, y_test).round(3)), '\n')

C = 0.001
training accuracy: 0.977
test accuracy: 0.977 

C = 0.01
training accuracy: 0.978
test accuracy: 0.979 

C = 0.1
training accuracy: 0.979
test accuracy: 0.979 

C = 1
training accuracy: 0.982
test accuracy: 0.988 

C = 10
training accuracy: 0.983
test accuracy: 0.99 



In [ ]:
# Cross-validation with tuning regularization in logistic regression

for i in cset:
    print('C =', i)
    log_reg = LogisticRegression(solver = 'lbfgs', 
                                 max_iter = 3000, 
                                 C = i)
    
    scores = cross_val_score(log_reg, 
                             df[xcols2], 
                             df['Selected?'], 
                             cv = 5)
    print(scores)
    print('Accuracy: %0.3f (+/- %0.3f)' % (scores.mean(), scores.std() * 2), '\n')

C = 0.001
[0.97709163 0.96111665 0.98404786 0.98105683 0.97507478]
Accuracy: 0.976 (+/- 0.016) 

C = 0.01
[0.9750996  0.96510469 0.98404786 0.98005982 0.97906281]
Accuracy: 0.977 (+/- 0.013) 

C = 0.1
[0.9750996  0.96211366 0.98305085 0.98504487 0.97706879]
Accuracy: 0.976 (+/- 0.016) 

C = 1
[0.97011952 0.96410768 0.98404786 0.9890329  0.98005982]
Accuracy: 0.977 (+/- 0.018) 

C = 10
[0.96912351 0.96610169 0.98404786 0.98803589 0.98404786]
Accuracy: 0.978 (+/- 0.018) 



In [ ]:

y_pred_test = clf.predict(X_test)
print('  Recall: {:.3f}'.format(recall_score(y_test, y_pred_test)))
print('  Precision: {:.3f}'.format(precision_score(y_test, y_pred_test)))
print('  F1 score: {:.3f}'.format(f1_score(y_test, y_pred_test)))
cm = confusion_matrix(y_test, y_pred_test)
tn, fp, fn, tp = cm.ravel()
specificity = tn / (tn+fp)
print('  Specificity: {:.3f}'.format(specificity))

  Recall: 0.946
  Precision: 0.814
  F1 score: 0.875
  Specificity: 0.992


In [ ]:
# Confusion matrix

cm = confusion_matrix(y_test, y_pred_test)
sample = np.array([['TN', 'FP'], ['FN', 'TP']])
print('CM key:\n', sample, '\n')
print('CM for test:\n', cm)

CM key:
 [['TN' 'FP']
 ['FN' 'TP']] 

CM for test:
 [[959   8]
 [  2  35]]


In [ ]:
scaler = MinMaxScaler()

X = df[xcols2]
scaled_X = scaler.fit_transform(X)
y = df['Selected?']

In [ ]:

scaled_X = scaler.fit_transform(X_train)

tuned_parameters = {'C': [0.1, 0.5, 1, 5, 10, 50, 100]}

grid = GridSearchCV(LogisticRegression(solver='liblinear'), tuned_parameters, 
                    cv = 3, scoring = 'accuracy')

grid.fit(scaled_X, y_train)

print('mean of accuracies:', grid.cv_results_['mean_test_score'])
print('std dev of accuracies:', grid.cv_results_['std_test_score'])

# print best parameter after tuning 
print('best parameters:', grid.best_params_) 

# store the best estimator (In this case it's 'C': 100)
best_logreg = grid.best_estimator_

mean of accuracies: [0.9680967  0.97208499 0.9745772  0.97856437 0.97981038 0.98030882
 0.97881331]
std dev of accuracies: [0.00461639 0.00372044 0.00339221 0.00519363 0.00521846 0.00554213
 0.00675528]
best parameters: {'C': 50}


In [ ]:
# Random forest
param_grid = {'n_estimators': [10, 100, 250], 'max_samples': [.25, .5, 1]} 


# instantiate grid search object
grid = GridSearchCV(RandomForestClassifier(), param_grid, cv = 3)

# fitting the model for grid search 
grid.fit(X, y)

# print parameters, mean, and standard deviation of scores by iteration
for z in range(0, len(grid.cv_results_['params'])):
    print('\nparams:', grid.cv_results_['params'][z])
    print('mean of accuracies:', grid.cv_results_['mean_test_score'][z])
    print('std dev of accuracies:', grid.cv_results_['std_test_score'][z])

# print best parameter after tuning 
print('\n***best parameters:', grid.best_params_)
print('best score:', grid.best_score_)

# store the best estimator (in this case it's 'max_samples': 0.5, 'n_estimators': 100)
best_rf = grid.best_estimator_


params: {'max_samples': 0.25, 'n_estimators': 10}
mean of accuracies: 0.9744816586921851
std dev of accuracies: 0.004376893620434658

params: {'max_samples': 0.25, 'n_estimators': 100}
mean of accuracies: 0.9772727272727272
std dev of accuracies: 0.005940855034191136

params: {'max_samples': 0.25, 'n_estimators': 250}
mean of accuracies: 0.9764752791068579
std dev of accuracies: 0.005826012438677226

params: {'max_samples': 0.5, 'n_estimators': 10}
mean of accuracies: 0.9730861244019139
std dev of accuracies: 0.0068366938594426885

params: {'max_samples': 0.5, 'n_estimators': 100}
mean of accuracies: 0.9774720893141945
std dev of accuracies: 0.006648059011376821

params: {'max_samples': 0.5, 'n_estimators': 250}
mean of accuracies: 0.9778708133971291
std dev of accuracies: 0.004809544689718825

params: {'max_samples': 1, 'n_estimators': 10}
mean of accuracies: 0.9533492822966507
std dev of accuracies: 0.0

params: {'max_samples': 1, 'n_estimators': 100}
mean of accuracies: 0.953349282

In [ ]:
mydata = grouped[(grouped['season'] == 2020)]
mydata = mydata.drop('Selected?', axis=1, errors='coerce')
mydata

PLAYER_ID  season        PLAYER_NAME  ... PriorASG  GP_y    GP_Per
254        2544    2020       LeBron James  ...       16    25  0.920000
264        2546    2020    Carmelo Anthony  ...       10    22  0.954545
390        2730    2020      Dwight Howard  ...        8    23  1.000000
429        2738    2020     Andre Iguodala  ...        1    21  0.857143
534      101108    2020         Chris Paul  ...       10    22  0.909091
...         ...     ...                ...  ...      ...   ...       ...
5010    1630269    2020       Keandre Cook  ...        0    24  0.041667
5011    1630270    2020       Xavier Sneed  ...        0    24  0.125000
5012    1630271    2020     Brodric Thomas  ...        0    22  0.318182
5013    1630273    2020  Freddie Gillespie  ...        0    23  0.043478
5014    1630278    2020         Ade Murkey  ...        0    22  0.045455

[541 rows x 34 columns]

In [ ]:
mydata.fillna(0, inplace = True)
mydata.replace([np.inf, -np.inf], np.nan, inplace=True)

X_test.replace([np.inf, -np.inf], np.nan, inplace=True)
X_test.fillna(0, inplace=True)
X_train.replace([np.inf, -np.inf], np.nan, inplace=True)
X_train.fillna(0, inplace=True)

In [ ]:
print(xcols)

['Minutes', 'FGM', 'FGA', 'FG3M', 'FG3A', 'FTM', 'FTA', 'OREB', 'DREB', 'AST', 'STL', 'BLK', 'TO', 'PF', 'PTS', 'PLUS_MINUS', 'PER', 'TSP', 'W_PCT', 'Jersey', '2KRank', 'LastASG?', 'PriorASG']


In [ ]:
# time to predict
pred = clf.predict(mydata[xcols2])
print(pred)

[1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 1 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 1 1
 1 0 0 1 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 1 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 0 0 0 0 1 0 0 0 0 0 0 0 0 1 0 1
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0
 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 1 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 

In [ ]:
# add predictions to df
mydata['All-Star'] = pred.tolist()
print(mydata)

      PLAYER_ID  season        PLAYER_NAME  ... GP_y    GP_Per  All-Star
254        2544    2020       LeBron James  ...   25  0.920000         1
264        2546    2020    Carmelo Anthony  ...   22  0.954545         0
390        2730    2020      Dwight Howard  ...   23  1.000000         0
429        2738    2020     Andre Iguodala  ...   21  0.857143         0
534      101108    2020         Chris Paul  ...   22  0.909091         0
...         ...     ...                ...  ...  ...       ...       ...
5010    1630269    2020       Keandre Cook  ...   24  0.041667         0
5011    1630270    2020       Xavier Sneed  ...   24  0.125000         0
5012    1630271    2020     Brodric Thomas  ...   22  0.318182         0
5013    1630273    2020  Freddie Gillespie  ...   23  0.043478         0
5014    1630278    2020         Ade Murkey  ...   22  0.045455         0

[541 rows x 35 columns]


In [ ]:
# see the predictions
mydata.loc[mydata['All-Star'] == 1]

PLAYER_ID  season            PLAYER_NAME  ... GP_y    GP_Per  All-Star
254        2544    2020           LeBron James  ...   25  0.920000         1
847      201142    2020           Kevin Durant  ...   24  0.750000         1
1060     201566    2020      Russell Westbrook  ...   19  0.631579         1
1316     201935    2020           James Harden  ...   22  0.818182         1
1342     201939    2020          Stephen Curry  ...   23  1.000000         1
1662     202331    2020            Paul George  ...   24  0.875000         1
1880     202681    2020           Kyrie Irving  ...   24  0.666667         1
1986     202695    2020          Kawhi Leonard  ...   24  0.833333         1
1995     202696    2020         Nikola Vucevic  ...   25  1.000000         1
2015     202699    2020          Tobias Harris  ...   23  0.869565         1
2071     202710    2020           Jimmy Butler  ...   21  0.333333         1
2200     203076    2020          Anthony Davis  ...   25  0.800000         1
2217     203078    2020           Bradley Beal  ...   19  0.894737         1
2240     203081    2020         Damian Lillard  ...   22  0.954545         1
2439     203114    2020        Khris Middleton  ...   22  1.000000         1
2845     203497    2020            Rudy Gobert  ...   23  1.000000         1
2907     203507    2020  Giannis Antetokounmpo  ...   22  0.909091         1
3082     203897    2020            Zach LaVine  ...   22  1.000000         1
3255     203944    2020          Julius Randle  ...   25  1.000000         1
3291     203954    2020            Joel Embiid  ...   23  0.739130         1
3362     203999    2020           Nikola Jokic  ...   23  1.000000         1
3566    1626164    2020           Devin Booker  ...   22  0.818182         1
3801    1627732    2020            Ben Simmons  ...   23  0.913043         1
3810    1627734    2020       Domantas Sabonis  ...   23  1.000000         1
3844    1627742    2020         Brandon Ingram  ...   20  1.000000         1
3910    1627759    2020           Jaylen Brown  ...   20  1.000000         1
3923    1627763    2020        Malcolm Brogdon  ...   23  1.000000         1
4159    1628369    2020           Jayson Tatum  ...   20  0.750000         1
4182    1628378    2020       Donovan Mitchell  ...   23  0.913043         1
4219    1628389    2020            Bam Adebayo  ...   21  0.904762         1
4587    1629027    2020             Trae Young  ...   23  0.956522         1
4593    1629029    2020            Luka Doncic  ...   23  0.956522         1
4769    1629627    2020        Zion Williamson  ...   20  0.950000         1

[33 rows x 35 columns]

Predicted 33 all stars. There were 24 in real life. Is there a way to limit to 24 in the model?

**Starters:**
LeBron James, Anthony Davis, Kawhi Leonard, Luka Doncic, James Harden,
Giannis Antetokounmpo, Joel Embiid, Pascal Siakam, Kemba Walker, Trae Young

**Reserves:**
Damian Lillard, Ben Simmons, Nikola Jokic, Jayson Tatum, Chris Paul, Russell Westbrook, Domantas Sabonis, Khris Middleton, Bam Adebayo, Rudy Gobert, Kyle Lowry, Brandon Ingram, Donovan Mitchell

**Starters that were left out:** Pascal Siakam, Kemba Walker

**Reserves that were left out:** Chris Paul, Kyle Lowry

**MODEL SHOWS 83% ACCURACY OVERALL**